service/matching/taskListManager.go

```diff
@@ -474,22 +578,55 @@ func (c *taskListManagerImpl) trySyncMatch(task *persistence.TaskInfo) (*persist
 	}
 	// Request from the point of view of Add(Activity|Decision)Task operation.
 	// But it is getTask result from the point of view of a poll operation.
-	request := &getTaskResult{task: task, C: make(chan *syncMatchResponse, 1)}
+	request := &getTaskResult{task: task, C: make(chan *syncMatchResponse, 1), syncMatch: true}
+
+	rsv := c.rateLimiter.Reserve()
+	if !rsv.OK() {
+		scope := metrics.MatchingTaskListMgrScope
+		c.metricsClient.IncCounter(scope, metrics.AddThrottleCounter)
+		return nil, fmt.Errorf("cannot add to tasklist, limit exceeded")
+	}
 	select {
-	case c.syncMatch <- request: // poller goroutine picked up the task
+	case c.tasksForPoll <- request: // poller goroutine picked up the task
 		r := <-request.C
 		return r.response, r.err
 	default: // no poller waiting for tasks
+		rsv.Cancel()
 		return nil, nil
 	}
 }
 
+func (c *taskListManagerImpl) deliverBufferTasksForPoll() {
+deliverBufferTasksLoop:
+	for {
+		err := c.rateLimiter.Wait(c.cancelCtx)
+		if err != nil {
+			if err == context.Canceled {
+				c.logger.Warn("Tasklist manager context is cancelled, shutting down")
+				break deliverBufferTasksLoop
+			}
+			c.logger.Warn("Unable to send tasks for poll, limit exceeded")
+			c.metricsClient.IncCounter(metrics.MatchingTaskListMgrScope, metrics.AddThrottleCounter)
+			continue
+		}
+		select {
+		case task, ok := <-c.taskBuffer:
+			if !ok { // Task list getTasks pump is shutdown
+				break deliverBufferTasksLoop
+			}
+			c.tasksForPoll <- &getTaskResult{task: task}
+		case <-c.deliverBufferShutdownCh:
```

Comment:
Looks like we have 2 separate ways to shutdown this channel.

Reply:
Spoke offline, this is to satisfy existing tests. There is a comment as well.